test/model/src/include/model/kv_update.h

```diff
@@ -0,0 +1,230 @@
+         */
+        bool
+        operator()(const kv_update *left, const kv_update *right) const noexcept
+        {
+            if (left == right)
+                return false;
+            if (left == nullptr)
+                return true;
+            if (right == nullptr)
+                return false;
+            return left->_timestamp < right->_timestamp;
+        }
+
+        /*
+         * kv_update::timestamp_comparator::operator() --
+         *     Compare the update to the given timestamp.
+         */
+        bool
+        operator()(const kv_update *left, timestamp_t timestamp) const noexcept
+        {
+            if (left == nullptr)
+                return true;
+            return left->_timestamp < timestamp;
+        }
+
+        /*
+         * kv_update::timestamp_comparator::operator() --
+         *     Compare the update to the given timestamp.
+         */
+        bool
+        operator()(timestamp_t timestamp, const kv_update *right) const noexcept
+        {
+            if (right == nullptr)
+                return false;
+            return timestamp < right->_timestamp;
+        }
+    };
+
+    /*
+     * kv_update::kv_update --
+     *     Create a new instance.
+     */
+    inline kv_update(const data_value &value, timestamp_t timestamp) noexcept
+        : _value(value), _timestamp(timestamp)
+    {
+    }
+
+    /*
+     * kv_update::operator== --
+     *     Compare to another instance.
+     */
+    inline bool
+    operator==(const kv_update &other) const noexcept
+    {
+        return _value == other._value && _timestamp == other._timestamp;
+    }
+
+    /*
+     * kv_update::operator!= --
+     *     Compare to another instance.
+     */
+    inline bool
+    operator!=(const kv_update &other) const noexcept
+    {
+        return !(*this == other);
+    }
+
+    /*
+     * kv_update::operator< --
+     *     Compare to another instance.
+     */
+    inline bool
+    operator<(const kv_update &other) const noexcept
+    {
+        if (_timestamp != other._timestamp)
+            return _timestamp < other._timestamp;
+        if (_value != other._value)
+            return _value < other._value;
+        return true;
+    }
+
+    /*
+     * kv_update::operator<= --
+     *     Compare to another instance.
+     */
+    inline bool
+    operator<=(const kv_update &other) const noexcept
+    {
+        return *this < other || *this == other;
+    }
+
+    /*
+     * kv_update::operator> --
+     *     Compare to another instance.
+     */
+    inline bool
+    operator>(const kv_update &other) const noexcept
+    {
+        return !(*this <= other);
+    }
+
+    /*
+     * kv_update::operator>= --
+     *     Compare to another instance.
+     */
+    inline bool
+    operator>=(const kv_update &other) const noexcept
+    {
+        return !(*this < other);
+    }
+
+    /*
+     * kv_update::value --
+     *     Get the value.
+     */
+    inline const data_value &
+    value() const noexcept
```

Comment:
This expose internal data to a different scope/timeline. Have you considered copying or smart pointer?

Reply:
This also feels like somewhat of a judgment call, but it's less clear to me.

On one hand, tying the lifetime of the returned reference to the table feels safe, because it solves the aforementioned problem with having to use a `const char*` pointer in WiredTiger API calls, the value does not change, and we are not removing any old updates from the table... yet.

At some point, I may have to add the ability to consolidate old updates to save memory, i.e., when I'll add the operation to advance the oldest timestamp. But even then, it may not be necessary to remove old updates as this can be simply modeled just by setting a global timestamp property—but we may still need to remove old operations to save memory. And it is at that point that existing references will become invalid. And to complicate things further, the operation to advance the oldest timestamp should be safe to run concurrently with other operations. So in that sense, returning a copy of the value or a smart pointer may make sense.

A possible advantage to switching to returning copies would be that I would be able to switch `kv_table_item`'s `std::deque<kv_update*>` to `std::deque<kv_update>`, which would simplify a few things—including that ugly timestamp/update comparator that has to now work on both references and pointers.

There is an open question about how I would design parsing workload info from debug logging, as it would be appealing to parse an update directly into `kv_update`, and whether there would be any resulting complications with C pointer lifetimes as I explained earlier in my response to the `kv_table::name()` comment. But hopefully not...

I've been kind hesitant with smart pointers as I'm not sure the overhead and the complicated data type is worth it, but as I think about this more, they do have some appeal.

Let me think about this a little more. (And if we do end up switching this to copying or smart pointers, I'd have to think about the access patterns a bit more to decide which of those options would make more sense.)

Reply:
I appreciate there is an engineering trade off here, and I don't have a good enough sense of the need to offer anything but very generic suggestions. I will take time to remind you of something I'm certain you already know, but in my experience I forget when breaking new ground: it's likely to start safe take the potential performance hit, and optimize backward than try to shore up fast but awkwardly intertwined interfaces.

Reply:
I ended up changing this to return a copy of `data_value`. `get()` is currently used in tests and verification, and at least for now the data values are small, so making a copy should not be a big issue. So I decided to take a bit of hit there instead of using shared pointers, which would make returning values faster, but at the cost of extra memory in the common case—as I expect typical model runs in the future to include large numbers of updates. I can revisit this and change it to shared pointers it becomes an issue later.

Reply:
Sounds good.